New a new  README to the Playground CLI (#2284)
WordPress/wordpress-playground: packages/playground/cli/README.md

to your unique WordPress setup. With the Playground CLI, you can use the following top-level commands:

-   **`server`**: (Default) Starts a local WordPress server.
-   **`run-blueprint`**: Executes a Blueprint file without starting a web server.
-   **`build-snapshot`**: Builds a ZIP snapshot of a WordPress site based on a Blueprint.

The `server` command supports the following optional arguments:

-   `--port=<port>`: The port number for the server to listen on. Defaults to 9400.
-   `--outfile`: When building, write to this output file.
-   `--wp=<version>`: The version of WordPress to use. Defaults to the latest.
-   `--autoMount`: Automatically mount the current directory (plugin, theme, wp-content, etc.).
-   `--mount=<mapping>`: Manually mount a directory. Format: /host/path:/vfs/path. It can be used multiple times.
-   `--mountBeforeInstall`: Mount a directory to the PHP runtime before installing WordPress. You can provide `--mountBeforeInstall` or `--mount-before-install` multiple times. Format: `/host/path:/vfs/path`.
-   `--mountDir`: Mount a directory to the PHP runtime. You can provide `--mountDir` or `--mount-dir` multiple times. Format: `"/host/path"` `"/vfs/path"`

> -   `--mountDirBeforeInstall`: Mount a directory to the PHP runtime before installing WordPress. You can provide `--mountDirBeforeInstall` or `--mount-dir-before-install` multiple times. Format: `"/host/path"` `"/vfs/path"`

Reviewer: ```suggestion
-   `--mountBeforeInstall`: Mount a directory to the PHP runtime before installing WordPress. You can provide this option multiple times. Format: `/host/path:/vfs/path`.
-   `--mountDir`: Mount a directory to the PHP runtime. You can provide this option multiple times. Format: `"/host/path"` `"/vfs/path"`
-   `--mountDirBeforeInstall`: Mount a directory to the PHP runtime before installing WordPress. You can provide this option multiple times. Format: `"/host/path"` `"/vfs/path"`
```

With the Playground CLI, you can use the `--help` to get some support about the available commands.

```bash
npx @wp-playground/cli --help
```

## Working with Blueprints

Blueprint is a JSON file where you can pre-define the initial state of your WordPress instance. It provides several functionalities, like installing plugins and themes, creating content, setting WordPress options, and executing steps.

Below is an example of a Blueprint that installs a plugin, logs the user in, and opens the new post editor.

```JSON
{
	"$schema": "https://playground.wordpress.net/blueprint-schema.json",
	"landingPage": "/wp-admin/post-new.php",
	"steps": [
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "gutenberg"
			},
			"options": {
				"activate": true
			}
		},
		{
			"step": "login",
			"username": "admin",
			"password": "password"
		}
	]
}
```

You can prototype and test your Blueprint in the online Playground editor.

To use a Blueprint, create a file (e.g., my-blueprint.json) and run the following command:

```bash
npx @wp-playground/cli server --blueprint=./my-blueprint.json
```

## Interoperability

> This CLI package is not just a helpful tool. It drives interoperability between the in-browser Playground, CLI packages, and the PHP Blueprints library. Once complete, it will reuse the same internals as the website at https://playground.wordpress.org, whether we're talking about running PHP code, executing Blueprints, building snapshots, serving requests, or maintaining

Reviewer: ```suggestion
This CLI package is not just a helpful tool. It drives interoperability between the in-browser Playground, CLI packages, and the PHP Blueprints library. Once complete, it will reuse the same internals as the website at https://playground.wordpress.net, whether we're talking about running PHP code, executing Blueprints, building snapshots, serving requests, or maintaining
```

# WordPress Playground CLI

> `@wp-playground/cli` streamlines the process of setting up a local WordPress environment for development and testing. It utilizes WordPress Playground to set up a new WordPress environment seamlessly. As the antecessor `wp-now`, you can switch between PHP and WordPress versions only with a flag.

Reviewer: ```suggestion
`@wp-playground/cli` streamlines the process of setting up a local WordPress environment for development and testing. It utilizes WordPress Playground to set up a new WordPress environment seamlessly. As its predecessor `wp-now`, you can switch between PHP and WordPress versions only with a flag.
```

> The requirement to run the Playground CLI is to have installed `Node.js 20.18.3` or superior, which you can find at [Node.js website](https://nodejs.org/en/download).

Reviewer: ```suggestion
The requirement to run the Playground CLI is to have installed `Node.js 20.18.3` or higher, which you can find at the [Node.js website](https://nodejs.org/en/download).
```

## Quickstart

> Running the playground CLI is as simple as accessing your plugin or theme directory and running the following command:

Reviewer: ```suggestion
Running the Playground CLI is as simple as going to your plugin or theme directory and running the following command:
```

```bash
cd my-plugin-or-theme-directory
npx @wp-playground/cli server --autoMount
```

> The flag `--autoMount` will figure out if the project folder is a Plugin or a Theme for you. For more custom scenarios, we can work with the following example using the flag `--mount`.

Reviewer: ```suggestion
The flag `--autoMount` will figure out if the project folder is a plugin or a theme for you. For more custom scenarios, we can work with the following example using the flag `--mount`.
```

## Quickstart

Running the Playground CLI is as simple as going to your plugin or theme directory and running the following command:

```bash
cd my-plugin-or-theme-directory
npx @wp-playground/cli server --auto-mount
```

The flag `--auto-mount` will figure out if the project folder is a plugin or a theme for you. For more advanced mounting options, see the [Mounting Local Directories](#mounting-local-directories) section.

## Usage

You don't have to install `@wp-playground/cli`, you can run it directly with `npx`. This is the recommended way to use the CLI and requires no permanent installation. To run a vanilla WordPress website, you can run the command:

```bash

> npx @wp-playground/cli server

Reviewer 2: nit: I'd talk about `@wp-playground/cli@latest` here, otherwise npx can sometimes run an outdated version from the local cache

```

### Choosing a WordPress Version

> By default, the CLI will load the latest stable version of WordPress and PHP. To specify your preferred versions, you can use the flag `--wp=<version>` and `--php=<version>`:

Reviewer 2: It loads PHP 8.0 by default. The reason being it's noticeably faster than 7.4 (lowest [recommended on wp.org](https://wordpress.org/about/requirements/)) and, when we set that version, WordPress was still struggling with PHP 8.1+ compat. Maybe it's time to bump it, but I don't think we should aim for the most recent PHP as a default – things often break on these new versions and core needs time to catch up.

> You don't have to install `@wp-playground/cli`, you can run directly with `npx`. This is the recommended way to use the CLI and requires no permanent installation. To run a vanilla WordPress website, you can run the command:

Reviewer: ```suggestion
You don't have to install `@wp-playground/cli`, you can run it directly with `npx`. This is the recommended way to use the CLI and requires no permanent installation. To run a vanilla WordPress website, you can run the command:
```

 npx @wp-playground/cli server --wp=6.8 --php=8.4
```

### Mounting local Directories

`@wp-playground/cli` operates by mounting your local project files into a virtualized WordPress environment. This allows you to work on your plugin or theme with a live WordPress instance without any complex setup. You can do this automatically or manually.

> For full control, you can manually mount a local directory to a specific path inside the virtual WordPress installation. For example, to mount your current project folder into the plugins directory, use the `--mount` flag:

Reviewer 2: Let's also mention `--mount-before-install=./my-local-site:/wordpress` as a way of creating the site in a local filesystem instead of in VFS.

> By default, the CLI will load the latest stable version from WordPress. To set up the preferred version, we can use the flags `--wp=<version>`. This flag will switch to different versions on the fly:

Reviewer: ```suggestion
By default, the CLI will load the latest stable version of WordPress. To specify your preferred version, you can use the flag `--wp=<version>`:

```bash
 npx @wp-playground/cli server --wp=6.8
```

### Mounting local Directories

`@wp-playground/cli` operates by mounting your local project files into a virtualized WordPress environment. You can do this automatically or manually.

### Automatic Mounting with --autoMount

> The `--autoMount` flag is the easiest way to get started. It inspects the current directory and automatically mounts it to the correct location in the virtual WordPress site. It supports the following directory types:

Reviewer: ```suggestion
The `--autoMount` flag is the easiest way to get started. It inspects the current directory and automatically mounts it to the correct location in the virtual WordPress site. These are the supported directory types and how they are detected:
```

> -   **Plugin Mode:** Presence of a PHP file with "Plugin Name:" in its header.

Reviewer: ```suggestion
-   **Plugin Mode**: Presence of a PHP file with "Plugin Name:" in its header.
```

-   **Theme Mode**: Presence of a style.css file with "Theme Name:" in its header.
-   **wp-content Mode**: Presence of plugins and themes subdirectories.
-   **WordPress Mode**: Presence of a complete WordPress installation. The directory will be mounted to the root `/wordpress` folder.

### Manual Mounting

For more control, you can manually specify with the flag `--mount`. For example, to mount a local theme directory into a virtual `wp-content/themes` directory, we use the following command:

```bash
npx @wp-playground/cli server --mount=/wordpress/wp-content/themes
```

## Command and Arguments

> Playground CLI is simple, configurable, and unopinionated. You can set it up

Reviewer: ```suggestion
Playground CLI is simple, configurable, and unopinionated. You can set it up according
```

to your unique WordPress setup. With the playground CLI, you can use the following top-level commands:

-   **`server`**: (Default) Starts a local WordPress server.
-   **`run-blueprint`**: Executes a Blueprint file without starting a web server.
-   **`build-snapshot`**: Builds a ZIP snapshot of a WordPress site based on a Blueprint.

The `server` command supports the following optional arguments:

-   `--port=<port>`: The port number for the server to listen on. Defaults to 9400.
-   `--outfile`: When building, write to this output file.
-   `--wp=<version>`: The version of WordPress to use. Defaults to the latest.
-   `--autoMount`: Automatically mount the current directory (plugin, theme, wp-content, etc.).
-   `--mount=<mapping>`: Manually mount a directory. Format: /host/path:/vfs/path. It can be used multiple times.

> -   `--mountBeforeInstall`: Mount a directory to the PHP runtime before installing WordPress. You can provide `--mount-before-install` multiple times. Format: `/host/path:/vfs/path`.

Reviewer: Which one is correct? This switches between CamelCase and snake-case

Author: I think we should keep consistency over the flags. If we are using `--mount-dir`, we should apply the same standard to `--auto-mount`

Author: Both are supported, I will include them.

> -   `--mountDir`: Mount a directory to the PHP runtime. You can provide `--mount-dir` multiple times. Format: `"/host/path"` `"/vfs/path"`

Reviewer: Also here: Which one is correct? This switches between CamelCase and snake-case

Author: I will double-check it, it came from the help message and I didn't pay attention to it:

```bash
  --mountBeforeInstall             Mount a directory to the PHP runtime before installing WordPress. You can provide --mount-before-install multiple times. Format: /host/path:/vfs/path                             [array]
  --mountDir                       Mount a directory to the PHP runtime. You can provide --mount-dir multiple times. Format: "/host/path" "/vfs/path"                                                                [array]
  --mountDirBeforeInstall          Mount a directory to the PHP runtime before installing WordPress. You can provide --mount-before-install multiple times. Format: "/host/path" "/vfs/path"                         [array]
 
  ```

	"steps": [
		{
			"step": "installPlugin",
			"pluginZipFile": {
				"resource": "wordpress.org/plugins",
				"slug": "gutenberg"
			},
			"options": {
				"activate": true
			}
		},
		{
			"step": "login",
			"username": "admin",
			"password": "password"
		}
	]
}
```

The example of a Blueprint above installs a plugin, logs the user in, and opens the new post editor. To learn more about Blueprints, please check the [documentation](https://wordpress.github.io/wordpress-playground/blueprints).

To use a Blueprint, create a file (e.g., my-blueprint.json) and run the following command:

```bash
npx @wp-playground/cli server --blueprint=./my-blueprint.json
```

## Programmatic Usage with JavaScript

The Playground CLI can be controlled programmatically from your JavaScript code using the `runCLI` function. This allows you to integrate all CLI functionalities directly into your development workflow, for example, end-to-end testing.

```JavaScript
import { runCLI } from "@wp-playground/cli";

const cliServer = await runCLI({
      command: "server"
});
const handler = cliServer.requestHandler;
const php = await handler.getPrimaryPhp();
```

## Interoperability

> This CLI package is not just a helpful tool. It drives interoperability between the in-browser Playground, CLI packages, and the PHP Blueprints library. Once complete, it will reuse the same internals as the website at https://playground.wordpress.net, whether we're talking about running PHP code, executing Blueprints, building snapshots, serving requests, or maintaining

Reviewer 2: We don't need this section anymore. We're mostly there and it was mostly written for contributors anyway

## Comparisons

### Laravel Valet

-   @wp-playground/cli handles the entire WordPress installation for you.
-   It works across all desktop platforms (Mac, Linux, Windows).
-   It does not set up custom host domains for you.
-   It allows you to switch WordPress and PHP versions on the fly.

### wp-env

-   @wp-playground/cli supports non-WordPress projects (any PHP/HTML files).
-   It does not require Docker.
-   It is faster to start up for quick tests and development.

> -   It does not include lifecycle scripts or a persistent MySQL database.

Reviewer: I think we should make the "It" more clear. Repeat Playground? Replace with `...`?

> -   Allows you to switch WordPress and PHP versions on the fly.

Reviewer 2: Not quite on the fly, you still need to stop the server and run another command.

> -   Does not include lifecycle scripts or a persistent MySQL database.

Reviewer 2: Let's mention it can still be used with a MySQL database, it just doesn't ship one

# WordPress Playground CLI

> `@wp-playground/cli` streamlines the process of setting up a local WordPress environment for development and testing. It utilizes WordPress Playground to set up a new WordPress environment seamlessly. As its predecessor `wp-now`, you can switch between PHP and WordPress versions only with a flag.

Reviewer 3: "As with its predecessor"

# Table of contents

-   [Requirements](#requirements)
-   [Quickstart](#quickstart)
-   [Usage](#usage)
-   [Working with Blueprints](#working-with-blueprints)
-   [How can I contribute?](#how-can-i-contribute)

> -

Reviewer 3: Minor thing, but this extra list item can be deleted.

-   `--port=<port>`: The port number for the server to listen on. Defaults to 9400.
-   `--outfile`: When building, write to this output file.
-   `--wp=<version>`: The version of WordPress to use. Defaults to the latest.
-   `--auto-mount`: Automatically mount the current directory (plugin, theme, wp-content, etc.).
-   `--mount=<mapping>`: Manually mount a directory (can be used multiple times). Format: /host/path:/vfs/path
-   `--mount-before-install`: Mount a directory to the PHP runtime before WordPress installation (can be used multiple times). Format: `"/host/path:/vfs/path"`.
-   `--mount-dir`: Mount a directory to the PHP runtime (can be used multiple times). Format: `"/host/path"` `"/vfs/path"`.
-   `--mount-dir-before-install`: Mount a directory before WordPress installation (can be used multiple times). Format: `"/host/path"` `"/vfs/path"`
-   `--blueprint=<path>`: The path to a JSON Blueprint file to execute.
-   `--blueprint-may-read-adjacent-files`: Consent flag: Allow "bundled" resources in a local blueprint to read files in the same directory as the blueprint file.
-   `--login`: Automatically log the user in as an administrator.

> -   `--skip-word-pressSetup`: Do not download or install WordPress. Useful if you are mounting a full WordPress directory.

Reviewer 3: Should this be [--skip-wordpress-setup](https://github.com/WordPress/wordpress-playground/blob/08c1fa7cbd6128e44e32750b2d69d3c2461a43fd/packages/playground/cli/src/run-cli.ts#L136C12-L136C32)?
